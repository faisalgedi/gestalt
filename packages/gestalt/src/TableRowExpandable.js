// @flow strict
import React, { type Node, useState } from 'react';
import cx from 'classnames';
import styles from './Table.css';
import Box from './Box.js';
import IconButton from './IconButton.js';

type Props = {|
  children: Node,
  hoverStyle?: 'gray' | 'none',
  expandedContents: Node,
|};

export default function TableRowExpandable(props: Props): Node {
  const [expanded, setExpanded] = useState(false);
  const hoverStyle = props.hoverStyle || 'gray';
  const cs = hoverStyle === 'gray' ? cx(styles.hoverShadeGray) : null;

  return (
    <>
      <tr className={cs}>
        <td>
          <IconButton
            accessibilityLabel=""
            icon={expanded ? 'arrow-up' : 'arrow-down'}
            iconColor="darkGray"
            onClick={() => setExpanded(!expanded)}
          />
        </td>
        {props.children}
      </tr>
      {expanded && (
        <tr>
          <td colSpan={React.Children.count(props.children) + 1}>
            <Box padding={6}>{props.expandedContents}</Box>
          </td>
        </tr>
      )}
    </>
  );
}
