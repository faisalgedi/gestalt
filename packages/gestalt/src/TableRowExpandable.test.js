// @flow strict
import React from 'react';
import renderer from 'react-test-renderer';
import TableRowExpandable from './TableRowExpandable.js';

test('renders correctly', () => {
  const tree = renderer
    .create(
      <TableRowExpandable expandedContents={<div>Expanded Contents</div>}>
        <div>row cells</div>
      </TableRowExpandable>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders correctly with explicit hover', () => {
  const tree = renderer
    .create(
      <TableRowExpandable
        hoverStyle="gray"
        expandedContents={<div>Expanded Contents</div>}
      >
        <div>row cells</div>
      </TableRowExpandable>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders correctly without hover', () => {
  const tree = renderer
    .create(
      <TableRowExpandable
        hoverStyle="none"
        expandedContents={<div>Expanded Contents</div>}
      >
        <div>row cells</div>
      </TableRowExpandable>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
