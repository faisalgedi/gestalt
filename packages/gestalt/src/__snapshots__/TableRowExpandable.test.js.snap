// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<tr
  className="hoverShadeGray"
>
  <td>
    <button
      aria-label=""
      className="button tapTransition enabled"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <div
        className="pog transparent"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      >
        <svg
          aria-hidden={true}
          aria-label=""
          className="icon darkGray iconBlock"
          height={18}
          role="img"
          viewBox="0 0 24 24"
          width={18}
        >
          <path
            d="test-file-stub"
          />
        </svg>
      </div>
    </button>
  </td>
  <div>
    row cells
  </div>
</tr>
`;

exports[`renders correctly with explicit hover 1`] = `
<tr
  className="hoverShadeGray"
>
  <td>
    <button
      aria-label=""
      className="button tapTransition enabled"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <div
        className="pog transparent"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      >
        <svg
          aria-hidden={true}
          aria-label=""
          className="icon darkGray iconBlock"
          height={18}
          role="img"
          viewBox="0 0 24 24"
          width={18}
        >
          <path
            d="test-file-stub"
          />
        </svg>
      </div>
    </button>
  </td>
  <div>
    row cells
  </div>
</tr>
`;

exports[`renders correctly without hover 1`] = `
<tr
  className={null}
>
  <td>
    <button
      aria-label=""
      className="button tapTransition enabled"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchCancel={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <div
        className="pog transparent"
        style={
          Object {
            "height": 40,
            "width": 40,
          }
        }
      >
        <svg
          aria-hidden={true}
          aria-label=""
          className="icon darkGray iconBlock"
          height={18}
          role="img"
          viewBox="0 0 24 24"
          width={18}
        >
          <path
            d="test-file-stub"
          />
        </svg>
      </div>
    </button>
  </td>
  <div>
    row cells
  </div>
</tr>
`;
